# Storytelling: An example of runtime optimization in Python
## by Abram Hindle < hindle1@ualberta.ca >

* A long time ago a composer named Edward Elgar wrote a postcard to his girl friend. In the bottom corner of the postcard was a cryptic message.
  * Dorabella Cipher from Wikipedia https://en.wikipedia.org/wiki/Dorabella_Cipher
* Many cryptographers have tried to decode it as English. 
   * We found that that decipherments of the characters do not match English distributions of characters. 
   * We found better evidence it matched music. So we decided to come up with decipherments that were musical.



![Dorabella Cipher](Dorabella-cipher-image.png)

Our idea is the direction and bumps are meaningful so we make a symbolic encoding where direction and bumps are encoded.

⇐2 ⇒3 ⇖2 ⇐3 ⇐1 ⇑2 ⇓1 ⇐3 ⇗1 ⇙2 ⇖3 ⇘2 ⇘1 ⇖1 ⇘2 ⇑3 ⇘2 ⇘2 ⇑2 ⇒3 ⇒3 ⇘2 ⇖1 ⇙1 ⇙2 ⇙1 ⇑1 ⇖3 ⇘3 

⇓1 ⇘2 ⇓1 ⇑2 ⇙1 ⇐3 ⇗1 ⇗2 ⇐3 ⇖2 ⇘2 ⇘2 ⇖2 ⇑2 ⇖1 ⇘1 ⇓1 ⇘2 ⇖3 ⇘2 ⇑2 ⇓2 ⇘3 ⇘1 ⇖1 ⇙1 ⇗1 ⇗1 ⇙1 ⇖3 ⇘3 

⇖2 ⇘3 ⇑2 ⇓2 ⇘3 ⇖2 ⇖1 ⇓2 ⇓3 ⇑1 ⇘3 ⇖2 ⇘2 ⇑2 ⇓2 ⇘1 ⇘3 ⇖1 ⇐3 ⇒3 ⇑1 ⇘3 ⇑2 ⇐3 ⇖1 ⇙1 ⇐3

Here's our program. 

* It generates a mapping of symbols to notes
* The mapping must be in key (e.g. C-minor)
* Converts a document of symbols into notes. 
* Writes out the decrypted document to disk

It's too slow.

# Rules for Optimization

1. Don't do it.
2. Only do it if you can measure the quality you want to optimize
3. Don't do it yet.
4. Only do it if you record your original performance.
5. Make sure you're optimizing what you want optimized
6. Record original performance
7. Investigate performance
8. Change system
9. Measure new performance
10. Compare
11. If not enough goto 6.

# Advice: Don't optimize till you measure!
# Advice: If you measure, measure more than once!

# Knuth's views on the subject 

From C2 Wiki http://wiki.c2.com/?PrematureOptimization
    
> /Premature optimization is the root of all evil/ -- DonaldKnuth
> 
> In DonaldKnuth's paper "Structured Programming With Goto Statements", he wrote:
> > "Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%." 

We're not engaging in premature optimization. We have a legitimate runtime problem.

In [1]:
# Make sure the output dir is available
!mkdir abram.gen.out

mkdir: cannot create directory ‘abram.gen.out’: File exists


In [3]:
# -*- coding: utf-8 -*-
# UTF8 hacks invoked.
import sys
#reload(sys)
#sys.setdefaultencoding('utf-8')
#reload(sys)
import codecs
#UTF8Writer = codecs.getwriter('utf8')
#sys.stdout = UTF8Writer(sys.stdout)
import util
import argparse
import inkey
import music21
import copy
import itertools
import re
import json
import time

n_dirs = 8
n_bumps = 3
directions = {
    "⇑":0,
    "⇗":1,
    "⇒":2,
    "⇘":3,
    "⇓":4,
    "⇙":5,
    "⇐":6,
    "⇖":7
}
inverse_directions = dict([(directions[key],key) for key in directions])
dir_string = "".join([inverse_directions[x] for x in range(0,n_dirs)])

def token_to_dir_tuple(token):
    return (directions[token[0:-1]], int(token[-1]))

def dir_tuple_to_string(token):
    return inverse_directions[token[0]]+str(token[1])

def load_transcript(filename):
    tokens = util.load_split_file(filename)
    return [token_to_dir_tuple(token) for token in tokens]

def rotate_list(l):
    return l[1:] + [l[0]]

def n_rotated_lists(l,n):
    out = range(0,n)
    out[0] = l
    for i in range(1,n):
        out[i] = rotate_list(out[i-1])
    return out

full_note      = 4.0
half_note      = 2.0
quarter_note   = 1.0
eighth_note    = 0.5
sixteenth_note = 0.25
note_durations = [full_note, half_note, quarter_note, eighth_note, sixteenth_note]

def make_note( midi21_pitch, duration ):
    new_note = music21.note.Note(midi21_pitch)
    new_note.duration = music21.duration.Duration(duration) # is this right
    return new_note

def pitches_of_key(mkey):
    m = re.match('([A-G][-#]?)(major|minor)',mkey)
    k = m.group(1)
    maj = m.group(2)
    our_key = music21.key.Key(k,maj)
    our_pitches = our_key.pitches
    return our_pitches

def generate_abram_song(song_tuples, mkey, note_mapping, bump_mapping, octave):
    """ returns a list of music21 notes """
    our_pitches = pitches_of_key(mkey)
    # print(our_pitches)
    # modify the key to the appropriate octave
    our_key_notes = [inkey.note_octave_mod(note, octave-4) for note in our_pitches]
    # now get the note mapping to notes
    our_notes = [our_key_notes[x] for x in note_mapping]
    # process the tuple
    # print song_tuples
    # print bump_mapping
    assert sum([len(t)==2 for t in song_tuples]) == len(song_tuples)
    # print [ (our_notes[n], bump_mapping[b-1]) for n,b in song_tuples ]
    return [ make_note(our_notes[n], bump_mapping[b-1]) for n,b in song_tuples ]

def note_to_str(note):
    return note.fullName + " " + str(note.duration) #music21.Note.fullName(note)
def song_to_txt(song):
    return "\n".join([note_to_str(note) for note in song])

def save_textfile(txt, filename):
    fd = file(filename,"w")
    fd.write(txt)
    fd.close()

def song_to_mid(song):
    stream = music21.stream.Stream()
    for elm in song:
        stream.append(elm)
    return stream
    # mf = music21.midi.translate.streamToMidiFile(stream)

def save_midifile(midi, midi_file):
    fp = midi.write('midi', fp=midi_file)
    
def strify(l):
    return [str(x) for x in l]
    
def generate_and_save_song(song_tuples,mkey, note_mapping, bump_mapping,octave):
    song = generate_abram_song(song_tuples, mkey, note_mapping, bump_mapping, octave)
    file_name = "abram.gen.out/" + mkey + ".O." + str(octave) + ".N." + ".".join("".join(strify(note_mapping))) + ".T." + "_".join(strify(bump_mapping))
    txt_file_name = file_name + ".txt"
    mid_file_name = file_name + ".mid"
    txt = song_to_txt(song)
    save_textfile(txt, txt_file_name)
    # print(song)
    mid = song_to_mid(song)
    save_midifile(mid, mid_file_name)

__grid_search_print__ = True
def grid_search(params,param_keys,song_tuples,f=None,maxtime=None):
    if maxtime is None:
        maxtime = float("inf")
    start = time.time()
    iters = 0
    for t in itertools.product(*[params[x] for x in param_keys]):
        if __grid_search_print__:
            print(t)
        f(song_tuples,*t)
        iters += 1
        if (time.time() - start > maxtime):
            print("Times up! %s" % iters)
            break
    return iters

def my_main(maxtime=10):
    filename = "abram-dirs.txt"
    song_tuples = load_transcript(filename)
    print(inkey.get_all_keys().keys())
    #print("\t".join([dir_tuple_to_string(x) for x in our_tokens]))
    params = {
        "keys":inkey.get_all_keys().keys(),
        # permutations of indices
        #  could do all permutations too
        "dir_notes":n_rotated_lists(range(0,n_dirs),n_dirs),
        # "dir_notes":list(itertools.permutations(range(0,n_dirs),n_dirs)) # if you want 40k combos
        # map bumps to duration
        "bump_durations":list(itertools.permutations(note_durations,n_bumps)),
        "octaves":range(3,6),
    }
    return grid_search(params,["keys","dir_notes","bump_durations","octaves"], 
                song_tuples, f=generate_and_save_song, maxtime=maxtime)

def generate_and_estimate_song(estimate,song_tuples,mkey, note_mapping, bump_mapping,octave):
    song = generate_abram_song(song_tuples, mkey, note_mapping, bump_mapping, octave)
    file_name = "abram.gen.out/" + mkey + ".O." + str(octave) + ".N." + ".".join("".join(strify(note_mapping))) + ".T." + "_".join(strify(bump_mapping))
    txt = song_to_text(song)
    return estimate(txt), file_name
    
def my_estimate_all(estimator, transcription_song_tuples):
    estimates = dict()
    def helper(estimator,song_tuples,mkey, note_mapping, bump_mapping,octave):
        estimate, name = generate_and_estimate_song(estimator,song_tuples,mkey, note_mapping, bump_mapping,octave)
        estimates[name] = estimate
        print((estimate, name))
    
    params = {
        "keys":inkey.get_all_keys().keys(),
        # permutations of indices
        #  could do all permutations too
        # "dir_notes":n_rotated_lists(range(0,n_dirs),n_dirs),
        "dir_notes":list(itertools.permutations(range(0,n_dirs),n_dirs)), # if you want 40k combos
        # map bumps to duration
        "bump_durations":list(itertools.permutations(note_durations,n_bumps)),
        "octaves":range(3,6),
    }
    grid_search(params,["keys","dir_notes","bump_durations","octaves"], song_tuples, f=helper)
    pickle.dump(estimates, file('abram-song-estimates.pkl','w'))    

def my_estimate(args):
    filename = args.transcript
    song_tuples = load_transcript(filename)
    my_estimate_all(None,song_tuples)
    
def test_rotate_list():
    assert rotate_list([0,1,2,3]) == [1,2,3,0]
    assert rotate_list([0,1,2,3,4]) == [1,2,3,4,0]
    assert rotate_list(range(0,100))[0:-1] == range(1,100)
    input  = [0,1,2]
    output = [[0,1,2],[1,2,0],[2,0,1]]
    assert json.dumps(n_rotated_lists(input, 3)) == json.dumps(output)    

def test_notes():
    note = music21.pitch.Pitch("C#4") 
    new_note = make_note( note, quarter_note )
    assert new_note.duration.type == 'quarter'
    assert new_note.pitch.fullName == 'C-sharp in octave 4'    
    song = song_to_mid([copy.deepcopy(new_note) for i in range(0,10)])
    notes = list(song.notesAndRests)
    for i in range(0,len(notes)):
        assert notes[i].duration.type == new_note.duration.type
        assert notes[i].pitch.fullName == new_note.pitch.fullName
    cmajor = pitches_of_key("Cmajor")
    assert [x.name for x in cmajor] == ["C","D","E","F","G","A","B","C"]
    
def my_tests(args):
    test_rotate_list()
    test_notes()
    


In [4]:
# run our main for 5 seconds
my_main(5)


['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 1.0), 3)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 1.0), 4)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 1.0), 5)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 0.5), 3)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 0.5), 4)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 0.5), 5)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 0.25), 3)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 0.25), 4)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 2.0, 0.25), 5)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 1.0, 2.0), 3)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 1.0, 2.0), 4)
('Emajor', [0, 1, 2, 3, 4, 5, 6, 7], (4.0, 1.0, 2.0),

74

In [5]:
# demo?
! xdg-open abram.gen.out/Emajor.O.4.N.0.1.2.3.4.5.6.7.T.1.0_0.5_2.0.mid

In [6]:
# Let's measure performance!
__grid_search_print__ = False # disable debugging
seconds = 1  # measure for 1 second
repeats = 10 # measure 10 times
def mean(x):
    return sum(x)/len(x)
perf = map(lambda x: my_main(seconds),range(0,repeats))
perf

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 16
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 18
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 18
['Emajor',

[16, 18, 18, 18, 18, 17, 16, 16, 17, 18]

In [7]:
print("Current Performance ~ %s iters per %s seconds " % (mean(perf), seconds))

Current Performance ~ 17 iters per 1 seconds 


4. Original performance? 
   * 14-18 songs per second? 
   * That's really very poor.
5. Make sure we are optimizing what want optimized? 
   * Time.
   * But at least we know what we want: More iters per second.
6. Record original performance
   * 14-18 song per second
7. Investigate performance
   * We're going to use a profiler
8. Change system
   * not until we profile!

At this point we either have a good idea of what is slow or we need some help to find out what.

Profilers instrument running code and take statistics about the number of calls and the 

Checkout https://docs.python.org/2/library/profile.html

In [9]:
import cProfile

In [10]:
# Instrument my program and run my_main for 4 seconds
cProfile.run('my_main(4)','cpout.txt')

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 42


## FYI Performance decreased due to instrumentation

In [11]:
import pstats

In [12]:
# Let's summarize the profiling run
p = pstats.Stats('cpout.txt')
p.strip_dirs().sort_stats(-1).print_stats() # order by name

Mon May 13 20:21:25 2019    cpout.txt

         5766906 function calls (5626374 primitive calls) in 4.069 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       42    0.002    0.000    0.100    0.002 <ipython-input-3-02a0ed9aa61e>:100(song_to_mid)
       42    0.000    0.000    2.055    0.049 <ipython-input-3-02a0ed9aa61e>:107(save_midifile)
       84    0.000    0.000    0.000    0.000 <ipython-input-3-02a0ed9aa61e>:110(strify)
       42    0.001    0.000    4.070    0.097 <ipython-input-3-02a0ed9aa61e>:113(generate_and_save_song)
        1    0.000    0.000    4.070    4.070 <ipython-input-3-02a0ed9aa61e>:125(grid_search)
        1    0.000    0.000    4.071    4.071 <ipython-input-3-02a0ed9aa61e>:140(my_main)
       87    0.000    0.000    0.000    0.000 <ipython-input-3-02a0ed9aa61e>:35(token_to_dir_tuple)
        1    0.000    0.000    0.000    0.000 <ipython-input-3-02a0ed9aa61e>:41(load_transcript)
        7    0.000

    52407    0.023    0.000    0.051    0.000 sites.py:94(_getAndUnwrapSite)
     3654    0.004    0.000    0.012    0.000 sites.py:952(removeById)
        6    0.000    0.000    0.000    0.000 socket.py:333(send)
    10836    0.005    0.000    0.007    0.000 sorting.py:112(__lt__)
    10962    0.007    0.000    0.013    0.000 sorting.py:96(__new__)
       42    0.000    0.000    0.001    0.000 spanner.py:618(__init__)
       42    0.000    0.000    0.000    0.000 spanner.py:649(__len__)
      294    0.000    0.000    0.000    0.000 streamStatus.py:179(client)
      252    0.001    0.000    0.001    0.000 streamStatus.py:72(__init__)
       42    0.000    0.000    0.001    0.000 streamStatus.py:87(__deepcopy__)
     7770    0.002    0.000    0.003    0.000 string.py:310(join)
      588    0.001    0.000    0.001    0.000 stringTools.py:98(getNumFromStr)
     1428    0.002    0.000    0.002    0.000 style.py:379(__init__)
       42    0.000    0.000    0.002    0.000 subConverters.py:80

<pstats.Stats instance at 0x7f4ad6bafbd8>

In [13]:
# Print by cumulative time
# That is the sum of time for all calls and child calls
# These are functions who call on others to do work
p.sort_stats('cumulative').print_stats(10)

Mon May 13 20:21:25 2019    cpout.txt

         5766906 function calls (5626374 primitive calls) in 4.069 seconds

   Ordered by: cumulative time
   List reduced from 421 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.071    4.071 <string>:1(<module>)
        1    0.000    0.000    4.071    4.071 <ipython-input-3-02a0ed9aa61e>:140(my_main)
        1    0.000    0.000    4.070    4.070 <ipython-input-3-02a0ed9aa61e>:125(grid_search)
       42    0.001    0.000    4.070    0.097 <ipython-input-3-02a0ed9aa61e>:113(generate_and_save_song)
       42    0.000    0.000    2.055    0.049 <ipython-input-3-02a0ed9aa61e>:107(save_midifile)
       42    0.000    0.000    2.055    0.049 __init__.py:246(write)
       42    0.000    0.000    2.055    0.049 base.py:2529(write)
       42    0.000    0.000    2.045    0.049 subConverters.py:921(write)
       42    0.000    0.000    1.989    0.047 translate.py:237(mu

<pstats.Stats instance at 0x7f4ad6bafbd8>

In [14]:
# total time is the time WITHIN the funciton but not subcalls
# these are the beefy functions where time is spent
p.sort_stats('tottime').print_stats(10)

Mon May 13 20:21:25 2019    cpout.txt

         5766906 function calls (5626374 primitive calls) in 4.069 seconds

   Ordered by: internal time
   List reduced from 421 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    32466    0.222    0.000    0.506    0.000 collections.py:50(__init__)
132552/3150    0.170    0.000    1.481    0.000 copy.py:145(deepcopy)
     3696    0.118    0.000    1.012    0.000 base.py:479(_deepcopySubclassable)
    19320    0.096    0.000    0.226    0.000 duration.py:1441(__init__)
     8484    0.095    0.000    0.138    0.000 pitch.py:2309(_setName)
    27972    0.091    0.000    0.592    0.000 base.py:409(__init__)
    32466    0.084    0.000    0.241    0.000 _abcoll.py:548(update)
    28644    0.071    0.000    0.221    0.000 pitch.py:2037(_getPs)
    31668    0.070    0.000    0.573    0.000 sites.py:167(__init__)
    21924    0.067    0.000    0.129    0.000 sites.py:330(add)




<pstats.Stats instance at 0x7f4ad6bafbd8>

# Looking at the results

Above, do you notice what library is taking up most of the top 6 places?

2 of those calls are to init, so that's object creation.

There's also a reference to deepcopy.

In [15]:
# this is the number of times a function is called
# speeding these up just a little bit can change
# their total time drasticly
p.sort_stats('ncalls').print_stats(20)

Mon May 13 20:21:25 2019    cpout.txt

         5766906 function calls (5626374 primitive calls) in 4.069 seconds

   Ordered by: call count
   List reduced from 421 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   748440    0.044    0.000    0.044    0.000 {id}
   249900    0.047    0.000    0.049    0.000 {getattr}
   219781    0.056    0.000    0.133    0.000 {isinstance}
   210126    0.023    0.000    0.023    0.000 {method 'append' of 'list' objects}
   201180    0.028    0.000    0.028    0.000 {method 'get' of 'dict' objects}
   148218    0.029    0.000    0.030    0.000 {setattr}
139776/139692    0.015    0.000    0.015    0.000 {len}
132552/3150    0.170    0.000    1.481    0.000 copy.py:145(deepcopy)
   104244    0.020    0.000    0.020    0.000 {round}
    95802    0.040    0.000    0.040    0.000 _weakrefset.py:70(__contains__)
    89997    0.034    0.000    0.051    0.000 weakrefTools.py:49(unwrapWeakref)
    88830 

<pstats.Stats instance at 0x7f4ad6bafbd8>

# Cost of creation?

I am concerned about these calls to music21

      968    0.122    0.000    1.007    0.001 base.py:479(_deepcopySubclassable)
     2222    0.101    0.000    0.147    0.000 pitch.py:2309(_setName)
     7326    0.095    0.000    0.594    0.000 base.py:409(__init__)

In [13]:
# here's a candidate function that creates lots of music21 objects.

def make_note( midi21_pitch, duration ):
    new_note = music21.note.Note(midi21_pitch)
    new_note.duration = music21.duration.Duration(duration) # is this right
    return new_note


# Problem: Makes too many objects that are for some reason very slow to create.
    
These are notes, with set duration. Luckily how I use these notes is immutable. I do not change them.

Therefore I can SHARE notes if they are already created. That is I can memoize the output of this function!

In [16]:
# keep the old code to really make it
def _make_note( midi21_pitch, duration ):
    new_note = music21.note.Note(midi21_pitch)
    new_note.duration = music21.duration.Duration(duration) # is this right
    return new_note

# a place to cache the created notes
__note_cache__ = dict()
# memoize
def make_note( midi21_pitch, duration ):
    # make a key
    note_key = str(midi21_pitch)+str(duration)
    # if the note isn't in the cache make it
    if not note_key in __note_cache__:
        __note_cache__[note_key] = _make_note(midi21_pitch, duration)
    return __note_cache__[note_key]

In [17]:
# let's make this more repeatable
def bench(seconds=1,repeats=10):
    perf = map(lambda x: my_main(seconds),range(0,repeats))
    print("Current Performance ~ %s iters per %s seconds " % (mean(perf), seconds))
    print(perf)
    return perf

In [18]:
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']


StreamException: the object (<music21.note.Note D#>, id()=139959408677072) is already found in this Stream (<music21.stream.Stream 0x7f4adcfabb90>, id()=139959511726992)

# Ok when I was doing this I didn't have this problem

The music21 library has changed, so I guess so should my presentation of my problem. *sigh*

## Performance is not stable between revisions!

In [19]:
def song_to_mid(song):
    stream = music21.stream.Stream()
    for elm in song:
        note = music21.note.Note()
        note.pitches = elm.pitches
        note.duration = elm.duration
        stream.append(note)
    return stream

In [20]:
# run it
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 17
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 16
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 18
['Emajor',

[17, 16, 18, 19, 18, 16, 16, 18, 19, 18]

It doesn't look like performance improved very much :(

In [21]:
# profile it
cProfile.run('my_main(4)','cpout2.txt')


['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 43


In [22]:
# investigate the profile
p = pstats.Stats('cpout2.txt')
p.sort_stats('tottime').print_stats(10)
p.sort_stats('cumtime').print_stats(20)


Mon May 13 20:29:08 2019    cpout2.txt

         5674184 function calls (5536752 primitive calls) in 4.059 seconds

   Ordered by: internal time
   List reduced from 426 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
136724/3225    0.172    0.000    1.450    0.000 /usr/lib/python2.7/copy.py:145(deepcopy)
    33249    0.147    0.000    0.492    0.000 /usr/lib/python2.7/collections.py:50(__init__)
    12427    0.141    0.000    0.204    0.000 /home/hindle1/.local/lib/python2.7/site-packages/music21/pitch.py:2309(_setName)
    33249    0.131    0.000    0.298    0.000 /usr/lib/python2.7/_abcoll.py:548(update)
     3784    0.120    0.000    0.953    0.000 /home/hindle1/.local/lib/python2.7/site-packages/music21/base.py:479(_deepcopySubclassable)
    28648    0.093    0.000    0.624    0.000 /home/hindle1/.local/lib/python2.7/site-packages/music21/base.py:409(__init__)
    88424    0.076    0.000    0.076    0.000 /home/hindle1/.local/

<pstats.Stats instance at 0x7f4ad6ae50e0>

I saw this:
    
    40    0.000    0.000    1.475    0.037 <ipython-input-20-50c167bfa7f6>:67(pitches_of_key)
    
Very few calls. The function doesn't do much but it does expensive things. So maybe we can memoize or refactor it?

In [23]:
__pitches_of_key__ = dict()
def _pitches_of_key(mkey):
    m = re.match('([A-G][-#]?)(major|minor)',mkey)
    k = m.group(1)
    maj = m.group(2)
    our_key = music21.key.Key(k,maj)
    our_pitches = our_key.pitches
    return our_pitches

# memoize
def pitches_of_key(mkey):
    if not mkey in __pitches_of_key__:
        __pitches_of_key__[mkey] = _pitches_of_key(mkey)
    return __pitches_of_key__[mkey]



In [24]:
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 25
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 25
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 29
['Emajor',

[25, 25, 29, 27, 29, 28, 29, 29, 28, 29]

Boom. Performance Improved! 14 to 23. Good. 33% faster

In [25]:
cProfile.run('my_main(4)','cpout3.txt')
p = pstats.Stats('cpout3.txt')
p.sort_stats('tottime').print_stats(10)
p.sort_stats('cumtime').print_stats(20)


['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 61
Mon May 13 20:31:19 2019    cpout3.txt

         5903389 function calls (5726605 primitive calls) in 4.046 seconds

   Ordered by: internal time
   List reduced from 332 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
171959/610    0.203    0.000    2.149    0.004 /usr/lib/python2.7/copy.py:145(deepcopy)
     5368    0.175    0.000    1.475    0.000 /home/hindle1/.local/lib/python2.7/site-packages/music21/base.py:479(_deepcopySubclassable)
    10614    0.125    0.000    0.184    0.000 /home/hindle1/.local/lib/python2.7/site-packages/music21/pitch.py:2309(_setName)
    53439    0.117    0.000    0.117    0.00

<pstats.Stats instance at 0x7f4ad6a8ffc8>

In [26]:
p.sort_stats('ncalls').print_stats(30)

Mon May 13 20:31:19 2019    cpout3.txt

         5903389 function calls (5726605 primitive calls) in 4.046 seconds

   Ordered by: call count
   List reduced from 332 to 30 due to restriction <30>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  1014514    0.061    0.000    0.061    0.000 {id}
   258846    0.029    0.000    0.029    0.000 {method 'append' of 'list' objects}
   253414    0.037    0.000    0.037    0.000 {method 'get' of 'dict' objects}
   249183    0.052    0.000    0.054    0.000 {getattr}
   189143    0.045    0.000    0.090    0.000 {isinstance}
171959/610    0.203    0.000    2.149    0.004 /usr/lib/python2.7/copy.py:145(deepcopy)
   153615    0.032    0.000    0.032    0.000 {setattr}
   129015    0.017    0.000    0.017    0.000 {method 'startswith' of 'str' objects}
   117431    0.038    0.000    0.038    0.000 /home/hindle1/.local/lib/python2.7/site-packages/music21/common/weakrefTools.py:18(wrapWeakref)
   106079    0.077    0.000    0

<pstats.Stats instance at 0x7f4ad6a8ffc8>

We're manipulating the objects alot with their octaves.

In [27]:
# memoize
def pitches_of_key(mkey):
    if not mkey in __pitches_of_key__:
        __pitches_of_key__[mkey] = _pitches_of_key(mkey)
    return __pitches_of_key__[mkey]

__pitches_octave_mod__ = dict()
def pitch_octave_mod( pitch, octave ):
    mkey = str(pitch)+str(octave)
    #mkey = (pitch,octave)
    if not mkey  in __pitches_of_key__:
        __pitches_octave_mod__[mkey] = inkey.note_octave_mod(pitch, octave)
    return __pitches_octave_mod__[mkey]



# memoize
def _get_pitches_of_key_with_octave(mkey, octave):
    our_pitches = pitches_of_key(mkey)
    our_key_notes = [pitch_octave_mod(note, octave-4) for note in our_pitches]
    return our_key_notes

# memoize
__key_and_octave__ = dict()
def get_pitches_of_key_with_octave(mkey, octave):
    our_key = "%s %s" % (mkey,str(octave))
    if not our_key in __key_and_octave__:
        __key_and_octave__[our_key] = _get_pitches_of_key_with_octave(mkey, octave)
    return __key_and_octave__[our_key]    

def generate_abram_song(song_tuples, mkey, note_mapping, bump_mapping, octave):
    """ returns a list of music21 notes """
    # our_pitches = pitches_of_key(mkey)
    # # print(our_pitches) 
    # # modify the key to the appropriate octave
    # our_key_notes = [pitch_octave_mod(note, octave-4) for note in our_pitches]
    our_key_notes = get_pitches_of_key_with_octave(mkey, octave)
    # now get the note mapping to notes
    our_notes = [our_key_notes[x] for x in note_mapping]
    # process the tuple
    # print song_tuples
    # print bump_mapping
    assert sum([len(t)==2 for t in song_tuples]) == len(song_tuples)
    # print [ (our_notes[n], bump_mapping[b-1]) for n,b in song_tuples ]
    return [ make_note(our_notes[n], bump_mapping[b-1]) for n,b in song_tuples ]



In [28]:
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 25
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 29
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 27
['Emajor',

[25, 29, 27, 28, 26, 25, 24, 29, 28, 29]

No real gain from that.

In [29]:
cProfile.run('my_main(4)','cpout4.txt')


['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 65


In [30]:
p = pstats.Stats('cpout4.txt')
# p.sort_stats('cumtime').print_stats(25)
p.strip_dirs().sort_stats('cumtime').print_stats(50)

Mon May 13 20:34:03 2019    cpout4.txt

         6250507 function calls (6064161 primitive calls) in 4.047 seconds

   Ordered by: cumulative time
   List reduced from 325 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.047    4.047 <string>:1(<module>)
        1    0.000    0.000    4.047    4.047 <ipython-input-3-02a0ed9aa61e>:140(my_main)
        1    0.000    0.000    4.046    4.046 <ipython-input-3-02a0ed9aa61e>:125(grid_search)
       65    0.001    0.000    4.045    0.062 <ipython-input-3-02a0ed9aa61e>:113(generate_and_save_song)
       65    0.000    0.000    3.244    0.050 <ipython-input-3-02a0ed9aa61e>:107(save_midifile)
       65    0.000    0.000    3.244    0.050 __init__.py:246(write)
       65    0.001    0.000    3.243    0.050 base.py:2529(write)
       65    0.001    0.000    3.227    0.050 subConverters.py:921(write)
       65    0.000    0.000    3.142    0.048 translate.py:237(m

<pstats.Stats instance at 0x7f4ad6730290>

# Awful Optimization Trick: Remove a layer of abstraction

Music21 objects are pretty slow.

       65    0.001    0.000    3.138    0.048 __init__.py:246(write)
       65    0.001    0.000    3.138    0.048 base.py:2529(write)
       65    0.001    0.000    3.121    0.048 subConverters.py:921(write)
       65    0.000    0.000    3.033    0.047 translate.py:237(music21ObjectToMidiFile)
       65    0.002    0.000    3.033    0.047 translate.py:1890(streamToMidiFile)
       65    0.005    0.000    3.031    0.047 translate.py:1680(streamHierarchyToMidiTracks)
       65    0.003    0.000    2.130    0.033 translate.py:1646(_prepareStreamForMidi)
    180691/130    0.202    0.000    2.021    0.016 copy.py:145(deepcopy)
       65    0.000    0.000    2.019    0.031 __init__.py:1455(__deepcopy__)
       65    0.012    0.000    1.860    0.029 __init__.py:1393(_deepcopySubclassable)

We're going to drop making music21 objects until we need them in the end.

We're going to convert notes to tuples instead of Music21 objects and make Music21 later.

In [31]:
# state is evil
__key_and_octave__ = dict()
__pitches_of_key__ = dict()
__pitches_octave_mod__ = dict()

# make a tuple
def _make_note( midi_pitch, duration ):
    return (midi_pitch, duration)
def make_note( midi_pitch, duration ):
    return _make_note(midi_pitch, duration)
def _get_pitches_of_key_with_octave(mkey, octave):
    our_pitches = pitches_of_key(mkey)
    our_key_notes = [pitch_octave_mod(note, octave-4) for note in our_pitches]
    # convert to midi
    return [p.midi for p in our_key_notes]
    # return our_key_notes
def note_to_midi(note):
    # if not note in __note_to_midi__:
    #     __note_to_midi__[note] = note.pitch.midi # supposedly this is slow
    # return __note_to_midi__[note]
    return int(note[0])

def song_to_txt(song):
    # return " ".join(["M{:03d}".format(note.midi) for note in song])
    return " ".join(["M{:03d}".format(note_to_midi(note)) for note in song])


In [32]:
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']


AttributeError: 'tuple' object has no attribute 'pitches'

In [33]:
def make_music21_note( midi21_pitch, duration ):
    new_note = music21.note.Note(midi21_pitch)
    new_note.duration = music21.duration.Duration(duration) # is this right    
    return new_note

def song_to_mid(song):
    stream = music21.stream.Stream()
    for elm in song:
        if isinstance(elm, music21.note.Note):
            #note = music21.note.Note()
            #note.pitches = elm.pitches
            #note.duration = elm.duration
            stream.append(elm)
        else:
            stream.append(make_music21_note(elm[0],elm[1]))
    return stream


In [34]:
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 24
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 20
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 24
['Emajor',

[24, 20, 24, 21, 25, 24, 26, 26, 24, 22]

In [35]:
cProfile.run('my_main(4)','cpout5.txt')


['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 55


In [36]:
p = pstats.Stats('cpout5.txt')
# p.sort_stats('cumtime').print_stats(25)
p.sort_stats('cumtime').print_stats(50)

Mon May 13 20:37:28 2019    cpout5.txt

         6108455 function calls (5882130 primitive calls) in 4.038 seconds

   Ordered by: cumulative time
   List reduced from 316 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.038    4.038 <string>:1(<module>)
        1    0.000    0.000    4.038    4.038 <ipython-input-3-02a0ed9aa61e>:140(my_main)
        1    0.001    0.001    4.038    4.038 <ipython-input-3-02a0ed9aa61e>:125(grid_search)
       55    0.001    0.000    4.037    0.073 <ipython-input-3-02a0ed9aa61e>:113(generate_and_save_song)
       55    0.000    0.000    3.385    0.062 <ipython-input-3-02a0ed9aa61e>:107(save_midifile)
       55    0.000    0.000    3.385    0.062 /home/hindle1/.local/lib/python2.7/site-packages/music21/stream/__init__.py:246(write)
       55    0.001    0.000    3.384    0.062 /home/hindle1/.local/lib/python2.7/site-packages/music21/base.py:2529(write)
       55    0.00

<pstats.Stats instance at 0x7f4ad6bdcfc8>

# It's still the midi output.

       55    0.000    0.000    3.354    0.061 <ipython-input-2-02a0ed9aa61e>:107(save_midifile)

Let's try it without without midi altogether.

# Imagine we asked our boss if we could just skip producing MIDI files?

If they said yes and changed the requirements we'd make huge gains.

In [37]:
__grid_search_print__ = False

def save_midifile(x,y):
    return None
def song_to_mid(x):
    return None

In [38]:
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 7455
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 8641
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 7938
['Em

[7455, 8641, 7938, 9379, 9412, 9424, 9503, 9410, 9497, 9201]

In [39]:
cProfile.run('my_main(4)','cpout6.txt')
p = pstats.Stats('cpout6.txt')
p.sort_stats('cumtime').print_stats(50)

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 20134
Mon May 13 20:38:37 2019    cpout6.txt

         9388012 function calls (9385254 primitive calls) in 4.001 seconds

   Ordered by: cumulative time
   List reduced from 226 to 50 due to restriction <50>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.001    4.001 <string>:1(<module>)
        1    0.000    0.000    4.001    4.001 <ipython-input-3-02a0ed9aa61e>:140(my_main)
        1    0.066    0.066    4.000    4.000 <ipython-input-3-02a0ed9aa61e>:125(grid_search)
    20134    0.091    0.000    3.931    0.000 <ipython-input-3-02a0ed9aa61e>:113(generate_and_save_song)
    20134    0.736    0.000    1.635

<pstats.Stats instance at 0x7f4ad6b46cf8>

  1808295    0.359    0.000    0.516    0.000 <ipython-input-29-d4e378dee91b>:9(make_note)
    
```make_note``` still is too much



In [40]:
# remove the function call to _make_note
def make_note( midi_pitch, duration ):
    return (midi_pitch, duration)
bench()
cProfile.run('my_main(4)','cpout7.txt')
p = pstats.Stats('cpout7.txt')
p.sort_stats('cumtime').print_stats(20)

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 8965
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 9630
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 9343
['Em

<pstats.Stats instance at 0x7f4ad6c21fc8>

# Mild improvement 8800 to 9300 iters.

Can we memoize further? 

In [41]:
# What about memozing it?
from collections import defaultdict
__make_note__ = defaultdict(dict)
def make_note( midi_pitch, duration ):
    if not duration in __make_note__[midi_pitch]:
        __make_note__[midi_pitch][duration] = (midi_pitch, duration)
    return __make_note__[midi_pitch][duration]

bench()
cProfile.run('my_main(4)','cpout8.txt')
p = pstats.Stats('cpout8.txt')
p.sort_stats('cumtime').print_stats(20)

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 8269
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 7962
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 8812
['Em

<pstats.Stats instance at 0x7f4ad6deed40>

# OK that didn't work, we're back to 8700

    2005872    0.377    0.000    0.377    0.000 <ipython-input-39-a77dbd047ed2>:4(make_note)
        
Yeah let's undo that change.

In [42]:
# nope memoize did not help
def make_note( midi_pitch, duration ):
    return (midi_pitch, duration)


Well that's neat. But maybe we should do something else? Like actually save midi files to disk?

# Ok but our boss didn't let us off the hook. We still need MIDI output.

* I don't want to change music21
* Maybe I can use another library to output midi?

In [43]:
import midi
TEMPO = 120
TICK = 60.0/(4*16.0*180)

def midi_on_off(when, pitch, duration):
    tick_on_when  = int((when)/TICK)
    tick_off_when = int((duration+when)/TICK)
    on  =  midi.NoteOnEvent(tick=tick_on_when,  pitch=pitch, velocity=100)
    off = midi.NoteOffEvent(tick=tick_off_when, pitch=pitch, velocity=100)
    return (when+duration, on, off)

def song_to_mid(song):
    # Instantiate a MIDI Pattern (contains a list of tracks)
    pattern = midi.Pattern()
    # Instantiate a MIDI Track (contains a list of MIDI events)
    track = midi.Track()
    # Append the track to the pattern
    pattern.append(track)
    # set tempo
    track.append(midi.SetTempoEvent(bpm=TEMPO))
    when = 0.0
    for elm in song:
        if isinstance(elm, music21.note.Note):
            (now, on, off) = midi_on_off(when, elm.midi, elm.duration)
        else:
            (now,on,off) = midi_on_off(when,elm[0],elm[1])
        track.append(on)
        track.append(off)
        when = 0.0 # don't need to keep track of time
    # Add the end of track event, append it to the track
    eot = midi.EndOfTrackEvent(tick=1)
    track.append(eot)
    return pattern

def save_midifile(midi_pattern, midi_file):
    midi.write_midifile(midi_file, midi_pattern)



In [44]:
test_midi = song_to_mid([(60,1.0),(64,1.0),(68,1.0),(60,0.25),(64,0.25),(68,0.25)])
print(test_midi)
save_midifile(test_midi,"test.mid")

midi.Pattern(format=1, resolution=220, tracks=\
[midi.Track(\
  [midi.SetTempoEvent(tick=0, data=[7, 161, 32]),
   midi.NoteOnEvent(tick=0, channel=0, data=[60, 100]),
   midi.NoteOffEvent(tick=192, channel=0, data=[60, 100]),
   midi.NoteOnEvent(tick=0, channel=0, data=[64, 100]),
   midi.NoteOffEvent(tick=192, channel=0, data=[64, 100]),
   midi.NoteOnEvent(tick=0, channel=0, data=[68, 100]),
   midi.NoteOffEvent(tick=192, channel=0, data=[68, 100]),
   midi.NoteOnEvent(tick=0, channel=0, data=[60, 100]),
   midi.NoteOffEvent(tick=48, channel=0, data=[60, 100]),
   midi.NoteOnEvent(tick=0, channel=0, data=[64, 100]),
   midi.NoteOffEvent(tick=48, channel=0, data=[64, 100]),
   midi.NoteOnEvent(tick=0, channel=0, data=[68, 100]),
   midi.NoteOffEvent(tick=48, channel=0, data=[68, 100]),
   midi.EndOfTrackEvent(tick=1, data=[])])])


In [45]:
! xdg-open test.mid

In [46]:
res = bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 981
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 945
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1003
['Emaj

In [47]:
cProfile.run('my_main(4)','cpout9.txt')
p = pstats.Stats('cpout9.txt')
p.sort_stats('cumtime').print_stats(20)

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 2306
Mon May 13 20:43:05 2019    cpout9.txt

         8689977 function calls in 4.001 seconds

   Ordered by: cumulative time
   List reduced from 70 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.001    4.001 <string>:1(<module>)
        1    0.000    0.000    4.001    4.001 <ipython-input-3-02a0ed9aa61e>:140(my_main)
        1    0.049    0.049    4.000    4.000 <ipython-input-3-02a0ed9aa61e>:125(grid_search)
     2306    0.017    0.000    3.951    0.002 <ipython-input-3-02a0ed9aa61e>:113(generate_and_save_song)
     2306    0.211    0.000    2.010    0.001 <ipython-input-43-

<pstats.Stats instance at 0x7f4ad6b21e18>

# So making MIDIs is still slow

     2332    0.206    0.000    2.012    0.001 <ipython-input-41-fa33ae1b567f>:12(song_to_mid)
   202884    0.347    0.000    1.711    0.000 <ipython-input-41-fa33ae1b567f>:5(midi_on_off)
     2332    0.040    0.000    1.510    0.001 <ipython-input-41-fa33ae1b567f>:35(save_midifile)
    
 

In [48]:
def _midi_on_off(when, pitch, duration):
    tick_on_when  = int((when)/TICK)
    tick_off_when = int((duration+when)/TICK)
    on  =  midi.NoteOnEvent(tick=tick_on_when,  pitch=pitch, velocity=100)
    off = midi.NoteOffEvent(tick=tick_off_when, pitch=pitch, velocity=100)
    return (when+duration, on, off)

__midi_on_off__ = dict()
def midi_on_off(when, pitch, duration):
    t = (when,pitch,duration)
    if not t in __midi_on_off__:
        __midi_on_off__[t] = _midi_on_off(when,pitch,duration)
    return __midi_on_off__[t]

In [49]:
bench()

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1726
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1565
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1476
['Em

[1726, 1565, 1476, 1816, 1585, 1565, 1852, 1598, 1647, 1912]

In [53]:
cProfile.run('my_main(4)','cpout10.txt')
p = pstats.Stats('cpout10.txt')
p.sort_stats('cumtime').print_stats(20)

['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 4206
Mon May 13 13:59:56 2019    cpout10.txt

         9263783 function calls in 4.003 seconds

   Ordered by: cumulative time
   List reduced from 71 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.003    4.003 <string>:1(<module>)
        1    0.000    0.000    4.003    4.003 <ipython-input-2-02a0ed9aa61e>:140(my_main)
        1    0.024    0.024    4.000    4.000 <ipython-input-2-02a0ed9aa61e>:125(grid_search)
     4206    0.026    0.000    3.975    0.001 <ipython-input-2-02a0ed9aa61e>:113(generate_and_save_song)
     4206    0.073    0.000    2.679    0.001 <ipython-input-41

<pstats.Stats instance at 0x7fabdc864200>

# Wow double performance. Great

> Current Performance ~ 1795 iters per 1 seconds 

Midi file saving still big but improved greatly!

In [50]:
# demo?
! xdg-open abram.gen.out/Emajor.O.4.N.7.0.1.2.3.4.5.6.T.1.0_0.5_2.0.mid

# Did we improve?

So from 14-20 iters per second to 1800 per second, doing the same work?

Good. About 80X or 6 orders of magnitude improvement.



# Conclusion

But at what cost?

* Less abstraction
* Less maintainble code
* Less use classes and types
* Less OO

But way faster.
Fundamentally I've simplified the problem and probably should do it in C or C++.

Other ways to improve speed:
* Do I/O in another thread. We're waiting a lot to save results
* Delegate disk I/O to another process
* Avoid disk I/O
* Use a more optimized method of saving
* Replace libraries
* use libs with faster bindings
* Remove a layer of abstraction
* Delegate work elsewhere
* Parallelism

# What is missing:

* Standardized benchmark
* Unit tests (you should always test for correctness)
* Statistic comparison of performance

# BONUS: But what if the improvement is subtle?

If you want to know if you have an improvement you should consider statistical tests:
    
## Student's T-Test

If the data is gaussian/normal/bell curved consider a Student's T-Test to see if your measurements come from the same normal distributions.

## Wilcoxon Rank Sum Test

If the data is not normal Wilcoxon Rank Sum Test can compare the distributions based on rank.

# Interpretting Statistical Tests

Choose an alpha, this is the probability threshold where you accept how many times you want to be wrong to chance. Alpha of 0.05 is common and for optimization is probably ok. Alpha of 0.1 is good for smaller number of runs. Essentially it's a protection against winning the lottery. Sometimes you get lucky.

If you p-value is less than your alpha the tests indicate evidence that your distributions are different. The difference can be negliable but they are statistically significantly different.

We'll use 0.05 as our alpha.

In [53]:
# you might want to run this to make sure the governor is the same
# by the time it runs the rest of the benchmarks
bench() 
res1 = bench()
res2 = bench()



['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1698
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1677
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1798
['Em

Times up! 1875
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1889
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Times up! 1899
['Emajor', 'B-minor', 'Gminor', 'E-minor', 'G-major', 'Aminor', 'D-major', 'Eminor', 'C-minor', 'Cmajor', 'C#major', 'Bminor', 'D-minor', 'Gmajor', 'F#minor', 'G-minor', 'E-major', 'Dminor', 'Fmajor', 'Cminor', 'C-major', 'B-major', 'Bmajor', 'C#minor', 'F#major', 'A-major', 'Dmajor', 'A-minor', 'Amajor', 'Fminor']
Time

In [54]:
import scipy.stats
import numpy
print("Means: %s %s" % (mean(res1),mean(res2)))
print("Comparison from same distribution")
wc = scipy.stats.wilcoxon( res1, res2 )
print(wc)
print(wc.pvalue <= 0.05)
print("Comparison from 10% worse performance")
wc = scipy.stats.wilcoxon( res1, 0.9*numpy.array(res2) )
print(wc)
print(wc.pvalue <= 0.05)
print("Comparison from same distribution w/ t-test")
tt = scipy.stats.ttest_ind( res1, res2 )
print(tt)
print(tt.pvalue <= 0.05)
print("Comparison from 10% worse performance w/ t-test")
tt = scipy.stats.ttest_ind( res1, 0.9*numpy.array(res2) )
print(tt)
print(tt.pvalue <= 0.05)


Means: 1872 1892
Comparison from same distribution
WilcoxonResult(statistic=13.0, pvalue=0.13941397332153205)
False
Comparison from 10% worse performance
WilcoxonResult(statistic=0.0, pvalue=0.005062032126267864)
True
Comparison from same distribution w/ t-test
Ttest_indResult(statistic=-1.9002561541348708, pvalue=0.0735388961386492)
False
Comparison from 10% worse performance w/ t-test
Ttest_indResult(statistic=16.225258871788704, pvalue=3.440783958287836e-12)
True


# BONUS: Do I have to time my own code?

No. ```timeit``` module can do it.

It can do weak reporting of timings. It is best for microbenchmarks.

In [55]:
import timeit
import music21
print(timeit.timeit('music21.note.Note("A#4")',
              setup='import music21',
              number=10000))
print(timeit.timeit('(70,0.25)',
              setup='',
              number=10000))

0.487522125244
0.000359058380127


# Other modules

```benchmark``` provides nicer reporting: https://pypi.org/project/benchmark/
        
Mike's "Python 101: An Intro to Benchmarking your code" https://www.blog.pythonlibrary.org/2016/05/24/python-101-an-intro-to-benchmarking-your-code/

In [ ]:
/* good CSS for presentation */
html body div {
        font-size: 20px;
        line-height: 1.2em;
}
.CodeMirror pre { 
        font-size: 20px;
        line-height: 1.2em;

}
